### ISSUE TITLE
Performance degradation

### ISSUE THREAD

**Issue by baoruxiao** [bonlime/keras-deeplab-v3-plus#8]
Hi,

I'm using the converted pretrain-weights to measure mIoU and observed 10% drop. Could you sure you measurement results, did you still get 84% mIoU on PASCAL using the transferred model and weights?

Thanks

**Comment by bonlime** [comment]
Hi, 
I didn't check the mIoU on PASCAL. The only thing I checked is that result of this model on the single image scale is identical to original model. 
I used original repo "deeplab_demo.ipynb" to check that, and got that all pixels are identical on all 3 images provided.  
To get semantic predictions add: 
`prediction = np.argmax(prediction,axis=-1)` where "predictions" are models output
According to original paper you should get 80.22% mIoU with single crop predictions, how many do you get?

**Comment by baoruxiao** [comment]
Hi bonlime, 

Thanks for the quick response. Actually I used your conversion to convert Tensorflow deeplabv3 trained on **CityScapes**, and measure the mIoU on CityScapes validation sets. They reported ~80% mIoU but I only got ~67%.

Is there something I need to be aware of if I use your conversion on CityScapes pretrained model?

I will compare the output of the same image, brilliant strategy.

Thanks

**Comment by bonlime** [comment]
How do you preprocess your image? The right way to do it is as follows:
1. resize image so that max(height,width) = 513
2. /127.5 - 1.
3. pad smaller dimension from one side with zeros so that img shape is (513,513)

**Comment by baoruxiao** [comment]
I used the same as you did, but cityscapes they use original cityscapes image size (1025, 2049) in evaluation, so what I did is only to do /127.5 - 1.

**Comment by bonlime** [comment]
--decoder_output_stride=4 
Just want to check, do you take this into account too? 
You need to take the output before bilinear resize, and accordingly resize true labels

**Comment by baoruxiao** [comment]
I didn't. Could you elaborate a little more why? 
Did you mean I need to only use the output of encoder, namely, xception_65?

**Comment by bonlime** [comment]
decoder_output_stride = 4 means they use only one x4 (or x2 in case of OS=8) bilinear upsample, but don't use the last one.  
you need to use "logits_semantic" layer output instead of original model output.  
you can define such new model like this  
`from keras import Model`  
`deeplab = Deeplabv3()`  
`model = Model(deeplab.input, deeplab.layers[-2].output)`

**Comment by baoruxiao** [comment]
The issue I found out is that:
Official Tensorflow evaluate mIoU by not considering the area labeled unknown (label_id = 255). see:
https://github.com/tensorflow/models/blob/5be198eca885fd967426ad5551c8374f1c3e889e/research/deeplab/eval.py#L126-L143

**Comment by bonlime** [comment]
Nice! So without unknown area, mIoU is the same?

**Comment by baoruxiao** [comment]
Yes, the same or just 1~2% lower than they reported depends on if implement the input preprocessing exactly as they do.

**Comment by lucasdavid** [comment]
I was also having some trouble related to this... I'm not sure what the issue is, but I managed to get the expected mIoU. I'll leave my comments in case it helps anyone:

My input comprises batches of 32 images of size 513x513. Images are scaled with the method `nearest` such that is largest dimension is 513 and the shortest size is padded with the pixel intensity 128. They are then processed with `f(x): return x/127.5 -1`.

I used the default `MeanIoU` metric implementation of Keras, with a small change to handle unknown label (-1, in my case) and a `y_pred` being a probability vector:

```py
class MeanIoU(tf.keras.metrics.MeanIoU):

  def __init__(
        self,
        num_classes,
        name=None,
        dtype=None,
        ignore_index: int = -1,
        sparse_predictions: bool = False,
    ):
    super().__init__(num_classes, name, dtype)

    self.ignore_index = ignore_index
    self.sparse_predictions = sparse_predictions

  def update_state(self, y_true, y_pred, sample_weight=None):
    y_true = tf.convert_to_tensor(y_true)
    y_pred = tf.convert_to_tensor(y_pred)

    if not self.sparse_predictions:
      y_pred = tf.argmax(y_pred, axis=-1)

    valid_mask = y_true != self.ignore_index
    y_true = y_true[valid_mask]
    y_pred = y_pred[valid_mask]

    return super().update_state(y_true, y_pred, sample_weight)

  def get_config(self):
    config = super().get_config()
    config.update({
      'ignore_index': self.ignore_index,
      'sparse_predictions': self.sparse_predictions,
    })

    return config
```

In this setting, I get mIoU = 66.62% over Pascal VOC 2012 validation set.

Then I fine-tuned the model for 20 epochs with `learning_rate=0.0001` (recommended value in deeplab's repo), nesterov momentum and `OS=8` over Pascal VOC 2012 augmented with Berkeley samples.
It converges to the lowest loss in the first epoch:
![loss over time](https://user-images.githubusercontent.com/1595030/160290024-61974914-2599-4005-8a8e-7bcb2ee77295.png)

And the final mIoU is 84.78% (value reported in the paper was 84.56%).
